clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -227,6 +231,66 @@ object GarbageCollector {
                                                spark,
                                                APIConfigurations(apiURL, accessKey, secretKey)
                                               ).withColumn("run_id", lit(runID))
-    expiredAddresses.write.partitionBy("run_id").mode(SaveMode.Append).parquet(addressesDFLocation)
+    expiredAddresses.write.partitionBy("run_id").mode(SaveMode.Append).parquet(addressesDFLocation) // TODO(Guys): consider changing to overwrite
   }
 }
+
+object S3BulkDeleter {
+
+  // collectByPartitionAndSize, for given column of type X return Lists[X] of size up to maxSize for rows within partition
+  def collectByPartitionAndSize(df: DataFrame, maxSize: Int, column: String) = {
+    val nRows = df.count()
+    val maxNRowsPartition = maxSize //make sure its a multiple of desired array length
+    val nPartitions = math.max(1, math.floor(nRows / maxNRowsPartition)).toInt
+    val repartitioned = df
+      .repartitionByRange(nPartitions, col(column).desc)
+      .withColumn("partitionId", spark_partition_id())
+
+    val w = Window.partitionBy(col("partitionId")).orderBy(column)
```

Comment:
Synced f2f: consider removing this part and using the partitions received by calling `repartitionByRange`. Those partitions may have more records than the target size, so this would need to be handled.